src/base/static/components/input-form/index.js

```diff
@@ -0,0 +1,551 @@
+      default:
+        fieldValue = autofillValue || fieldConfig.default_value || "";
+        break;
+    }
+
+    const fieldObj = ImmutableMap()
+      .set(constants.FIELD_STATE_VALUE_KEY, fieldValue)
+      .set(constants.FIELD_STATE_FIELD_TYPE_KEY, fieldConfig.type)
+      .set(
+        constants.FIELD_STATE_VALIDITY_KEY,
+        this.getValidator(fieldConfig).validate(fieldValue)
+      )
+      .set(
+        constants.FIELD_STATE_VALIDITY_MESSAGE_KEY,
+        this.getValidator(fieldConfig).message
+      );
+
+    return fieldObj;
+  }
+
+  resolveField(fieldConfig) {
+    if (fieldConfig.type === constants.COMMON_FORM_ELEMENT_TYPENAME) {
+      return this.props.placeConfig.common_form_elements[fieldConfig.name];
+    } else {
+      return fieldConfig;
+    }
+  }
+
+  onSubmit(evt) {
+    // TOOD: Enter button submission is still having an effect...
+
+    evt.preventDefault();
+    Util.log("USER", "new-place", "submit-place-btn-click");
+
+    // Validate the form.
+    const newValidationErrors = new Set();
+    let isValid = true;
+    this.state.fields.forEach((value, name) => {
+      if (!value.get(constants.FIELD_STATE_VALIDITY_KEY)) {
+        newValidationErrors.add(
+          value.get(constants.FIELD_STATE_VALIDITY_MESSAGE_KEY)
+        );
+        isValid = false;
+      }
+    });
+
+    if (isValid) {
+      this.saveModel();
+    } else {
+      this.setState({
+        formValidationErrors: newValidationErrors,
+        showValidityStatus: true,
+      });
+      this.scrollTo(document.querySelector(this.props.container), 0, 300);
+    }
+  }
+
+  // Due to https://stackoverflow.com/questions/8917921/cross-browser-javascript-not-jquery-scroll-to-top-animation
+  scrollTo(elt, to, duration) {
+    const difference = to - elt.scrollTop;
+    const perTick = difference / duration;
+    setTimeout(() => {
+      elt.scrollTop = elt.scrollTop + perTick;
+      if (elt.scrollTop === to) return;
+      this.scrollTo(elt, to, duration - 10);
+    }, 10);
+  }
+
+  saveModel() {
+    // TODO: this state should disable individual fields as well (?), not just
+    //       the submit button.
+    this.setState({
+      isFormSubmitting: true,
+    });
+
+    const selectedCategoryConfig = this.getCategoryConfig(
+      this.state.selectedCategory
+    );
+    let collection = this.props.places[selectedCategoryConfig.dataset];
+
+    collection.add({
+      location_type: selectedCategoryConfig.category,
+      datasetSlug: this.props.mapConfig.layers.find(
+        layer => selectedCategoryConfig.dataset === layer.id
+      ).slug,
+      datasetId: selectedCategoryConfig.dataset,
+      showMetadata: selectedCategoryConfig.showMetadata,
+    });
+    const model = collection.at(collection.length - 1);
+    let attrs = Object.assign(
+      {},
+      this.state.fields
+        .filter(
+          val =>
+            val.get(constants.FIELD_STATE_FIELD_TYPE_KEY) !==
+              constants.MAP_DRAWING_TOOLBAR_TYPENAME &&
```

Comment:
BTW this is more logic that shouldn't be in InputForm.

And wrapping this in Object.assign is redundant: https://facebook.github.io/immutable-js/docs/#/Map/toJS

Reply:
I'd argue this is appropriate logic for the form, because although the drawing toolbar and attachment field components are conceived of as fields, their output is handled differently from other fields: geometry is saved in a special field in the Django model, and attachments are an entirely separate model. This special handling is dictated by the back-end data modeling, of which the form is basically an extension. So it makes sense to me that special handling here would be at the form and not the field level.